### DISCUSSION TITLE
Is there a way to do authorization decisions in repsecxt to the payload of a message?

### DISCUSSION THREAD

**Discussion by raphaelahrens** [rabbitmq/rabbitmq-server#7582] [Questions] [answered]
I wanted to write a plugin for RabbitMQ which can make authorizations decisions in with  a label which is part of the payload of the message.
The idea was to check an incoming message for an exchange and set the label of the exchange to the label of the highest message.

I first though that by writing an AutZ-Plugin this should be simple, just unpack the payload in the check_*_access functions and get the label.

But to my surprise the Payload/Msg is not part of the AuthzContext .
I then looked into other ways to intercept the handling of the messages to get the payload and found Channel interceptors, but they only work for AMQP and not MQTT, since the MQTT plugin does not call the channel_interceptor.

Looking into the code of the MQTT-Plugin I don't see why it was not possible to include the Msg into the AuthZContext

https://github.com/rabbitmq/rabbitmq-server/blob/9b33e01a555476c63f3dfa863e367a3ea5b64160/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1527-L1534

Is there a reason for this decision?
Is there a another way than the channel interceptors to access the payload?

Other reason for doing checks on the payload

*  length restrictions
* signatures  checks

**Comment by michaelklishin** [comment, marked as answer]
There is no. RabbitMQ treats message payload as an opaque sequence of bytes. With the exception of dead lettering, the message time stamp and node stamp plugins, RabbitMQ does not mess with messages. Even then it does not touch the payload.

Authorization steps happen in the context of a protocol operation but message bodies are not present or involved at those moments.
